odl/contrib/fom/fom.py

```diff
@@ -0,0 +1,434 @@
+# Copyright 2014-2017 The ODL contributors
+#
+# This file is part of ODL.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at https://mozilla.org/MPL/2.0/.
+"""
+Implementation of Figure of Merits (FOMs) for comparing reconstructions with
+a given reference.
+"""
+
+__all__ = ('mean_square_error', 'mean_absolute_error', 'mean_density_value',
+           'density_standard_deviation', 'density_range', 'blurring',
+           'false_structures')
+
+
+def mean_square_error(test_data, ref_data, mask=None):
+    """Mean square error-based FOM between test_data and ref_data.
+
+    Evaluates mean square error FOM between input (test_data) and
+    reference data (ref_data), by measuring consistency using the L2 -norm.
+    Allows for a masking of the two spaces (mask).
+
+    Notes
+    ----------
+    The FOM evaluates
+
+    .. math::
+        1 - \\frac{\| f - g \|^2_2}{\| f \|^2_2 + \| g \|^2_2}.
+
+    The FOM takes its values in [0 , 1], with higher correspondance
+    at higher FOM value.
+
+    Parameters
+    ----------
+    test_data : `LinearSpace`
+        Input data or reconstruction.
+    ref_data : `LinearSpace`
+        Reference data to compare 'reco' to.
+    mask : `LinearSpace`
+        Binary mask to define ROI in which FOM evaluation is performed.
+
+    Returns
+    -------
+    fom : scalar
+        Scalar (float) indicating perfect correspondance (1) to no
+        correspondance (0)
+    """
+
+    import odl
+
+    l2_normSquared = odl.solvers.L2NormSquared(test_data.space)
+
+    if mask:
+        test_data = test_data * mask
+        ref_data = ref_data * mask
+
+    diff = test_data - ref_data
+    fom = 1.0 - l2_normSquared(diff) / (l2_normSquared(test_data) +
+                                        l2_normSquared(ref_data))
+    return fom
+
+
+def mean_absolute_error(test_data, ref_data, mask=None):
+    """Mean absolute error-based FOM between test_data and ref_data.
+
+    Evaluates mean absolute error FOM between input (test_data) and
+    reference data (ref_data), by measuring consistency using the L1-norm.
+    Allows for a masking of the two spaces (mask).
+
+    Notes
+    ----------
+    The FOM evaluates
+
+    .. math::
+        1 - \\frac{\| f - g \|_1}{\| f \|_1 + \| g \|_1}.
```

Comment:
I agree that this might be a somewhat weird. Perhaps one could do |f-g|_1/(2* max( |f|_1, |g|_1) ) instead, but I am not sure that this is better. Regarding references we have loosely been following section 2.4 in [this article](http://www.sciencedirect.com/science/article/pii/S1047847701943382). I have not found any reference for normalized FOMs unfortunately.